### PR TITLE
Verify types in `TypedData` in line with SNIP-12

### PR DESCRIPTION
**Pull request by franciszekjob**
Relates #1353 

## Introduced changes

Ensure type verification logic in line with [guidelines](https://github.com/starknet-io/SNIPs/pull/77/files). Restrictions related to `enum` basic type will be added separately in next PRs.

- Disallow empty names, names with `*`, dangling types
- Require domain type to be present in types
- Add tests for types redefinition, missing domain type, dangling type, missing dependency

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: starknet_py/utils/typed_data_test.py [hunk 1]
```diff
@@ -180,13 +180,6 @@ def test_message_hash(example, account_address, msg_hash):
     revision=Revision.V1,
 )
 
-domain_object_v1 = {
-    "name": "DomainV1",
-    "version": "1",
-    "chainId": "1234",
-    "revision": 1
-}
-
```
Comment (franciszekjob) [inline starknet_py/utils/typed_data_test.py:189]: Deleted, because it wasn't used anywhere.

File: starknet_py/utils/typed_data.py [hunk 2]
```diff
@@ -164,12 +168,35 @@ def _encode_data(self, type_name: str, data: dict) -> List[int]:
...
 
+        referenced_types = [
+            parameter for type_name in self.types for parameter in self.types[type_name]
+        ]
+        referenced_types = [
+            ref_type.contains
+            if ref_type.contains is not None
+            else strip_pointer(ref_type.type)
+            for ref_type in referenced_types
+        ] + [self.domain.separator_name, self.primary_type]
```
Comment (ddoktorski) [inline starknet_py/utils/typed_data.py:188]: nit: I think it does the same thing, but without duplicates and readability seems to be slightly better.
```suggestion
        referenced_types = {
            ref_type.contains
            if ref_type.contains is not None
            else strip_pointer(ref_type.type)
            for type_name in self.types
            for ref_type in self.types[type_name]
        }

        referenced_types.update([self.domain.separator_name, self.primary_type])
```

File: starknet_py/utils/typed_data.py [hunk 3]
```diff
@@ -164,12 +168,35 @@ def _encode_data(self, type_name: str, data: dict) -> List[int]:
...
+
+        for type_name in self.types:
+            if not type_name:
+                raise ValueError("Type names cannot be empty.")
+            if is_pointer(type_name):
+                raise ValueError(f"Type names cannot end in *. {type_name} was found.")
+            if type_name not in referenced_types:
+                raise ValueError(
+                    f"Dangling types are not allowed. Unreferenced type {type_name} was found."
```
Comment (ddoktorski) [inline starknet_py/utils/typed_data.py:197]: Can you explain why it is actually not allowed?

File: starknet_py/utils/typed_data_test.py [hunk 4]
```diff
@@ -197,15 +197,75 @@ def _make_typed_data(included_type: str, revision: Revision):
...
+    ],
+)
+def test_types_redefinition(included_type: str, revision: Revision):
     with pytest.raises(ValueError, match=f"Reserved type name: {included_type}"):
-        _make_typed_data(included_type, Revision.V1)
+        _make_typed_data(included_type, revision)
+
+
+def test_custom_type_definition():
+    _make_typed_data("myType", Revision.V0)
```
Comment (ddoktorski) [inline starknet_py/utils/typed_data_test.py:226]: What's the purpose of this test?

File: starknet_py/utils/typed_data.py [hunk 5]
```diff
@@ -132,6 +132,10 @@ def _encode_value(
             hashes = [self._encode_value(type_name, val) for val in value]
             return compute_hash_on_elements(hashes)
 
+        basic_types = [bt.value for bt in BasicType]
```
[the remark this answers is not in the document]
Comment (franciszekjob) [reply starknet_py/utils/typed_data.py]: Sure, I've added `_get_basic_types_values()` which retrieves all values from `BasicType`.

File: starknet_py/utils/typed_data_test.py (hunk 4 above)
Comment (franciszekjob) [reply starknet_py/utils/typed_data_test.py:226]: It tests whether it's possible to create type data with custom type as primary one. In Swift SDK there is a [similar test](https://github.com/software-mansion/starknet.swift/pull/170/files#diff-c38c472f61f02b67b364e318b6fa8fed796f64c1bc5bf21a2b065b9d55bcf442R92).

File: starknet_py/utils/typed_data.py (hunk 3 above)
Comment (franciszekjob) [reply starknet_py/utils/typed_data.py:195]: It's one of the rules of [types identification](https://github.com/starknet-io/SNIPs/pull/77/files#diff-d594c362a3c260ea39dacaca55a6f02927013cebd19a5c8d202117ff9147ff14R126)
Here's a link to the [subsection about no dangling types](https://github.com/starknet-io/SNIPs/pull/77/files#diff-d594c362a3c260ea39dacaca55a6f02927013cebd19a5c8d202117ff9147ff14R126)

**Review by ddoktorski — APPROVED** [review]

**Pull request merged**